hack/cnbi-package-list.yaml

```diff
@@ -0,0 +1,169 @@
+      runAfter:
+        - build-image
+      params:
+        - name: SCRIPT
+          value: |
+            cat <<EOM | oc apply -f -
+            ---
+            kind: ImageStream
+            apiVersion: image.openshift.io/v1
+            metadata:
+              annotations:
+                opendatahub.io/notebook-image-name: $(context.pipelineRun.name)
+              name: $(context.pipelineRun.name)
+              namespace: $(context.pipelineRun.namespace)
+              labels:
+                opendatahub.io/notebook-image: 'true'
+            spec:
+              lookupPolicy:
+                local: true
+              tags:
+                - name: latest
+            EOM
+---
+apiVersion: tekton.dev/v1beta1
+kind: Task
+metadata:
+  name: buildah-requirements
+spec:
+  params:
+    - name: BASE_IMAGE
+      type: string
+    - description: Reference of the image buildah will produce.
+      name: IMAGE
+      type: string
+    - default: >-
+        registry.redhat.io/rhel8/buildah@sha256:6a68ece207bc5fd8db2dd5cc2d0b53136236fb5178eb5b71eebe5d07a3c33d13
+      description: The location of the buildah builder image.
+      name: BUILDER_IMAGE
+      type: string
+    - default: vfs
+      description: Set buildah storage driver
+      name: STORAGE_DRIVER
+      type: string
+    - default: "true"
+      description: >-
+        Verify the TLS on the registry endpoint (for push/pull to a non-TLS
+        registry)
+      name: TLSVERIFY
+      type: string
+    - default: oci
+      description: "The format of the built container, oci or docker"
+      name: FORMAT
+      type: string
+  workspaces:
+    - name: requirements
+      readonly: true
+      description: the list of pinned python package we install in the produced image
+  volumes:
+    - name: /var/lib/containers
+      emptyDir: {}
+  steps:
+    - name: build-image
+      image: $(params.BUILDER_IMAGE)
+      env:
+        - name: STORAGE_DRIVER
+          value: $(params.STORAGE_DRIVER)
+        - name: BUILDAH_FORMAT
+          value: $(params.FORMAT)
+      script:
+        container = $(buildah from --tls-verify=$(params.TLS_VERIFY) docker://$(params.BASE_IMAGE))
+        buildah copy $container $(workspaces.requirements.path)/requirements-pinned.txt /tmp/
+        buildah run $container -- /usr/bin/pip install -r /tmp/requirements-pinned.txt
+        buildah commit $container $(params.IMAGE)
+        buildah push --tls-verify=$(params.TLS_VERIFY) $(params.IMAGE)
+      volumeMounts:
+        - name: containers
+          mountPath: /var/lib/containers
+---
+apiVersion: tekton.dev/v1beta1
+kind: Task
+metadata:
+  name: get-package-versions-list
+spec:
+  volumes:
+    - name: unpinned-packages
+      emptyDir:
+        medium: memory
+  workspaces:
+    - name: pinned-packages
+  params:
+    - name: CNBI_PACKAGE_LIST
+      type: string
+  steps:
+    - image: image-registry.openshift-image-registry.svc:5000/openshift/cli:latest
+      script: |
```

Comment:
Would the packagelist be provided as params to pipeline ?
if not, then we should at least have the operator to create a confimap of package list
getting package list from cnbi resource, doesn't seem be the right way.

Reply:
> Would the packagelist be provided as params to pipeline ?

Not in the current version of this PR. The only param is the name of the cnbi resources, and retrieving data from it is made entirely from the pipeline.
Both are possible, though using a param of array type array is a bit troublesome due to how tekton handle array paremeters expansion. (it can't be used in a `script`, see https://tekton.dev/docs/pipelines/tasks/#substituting-array-parameters)

> if not, then we should at least have the operator to create a confimap of package list getting package list from cnbi resource, doesn't seem be the right way.

Hum, why is that ? That just adds a step (cnbi -> configmap -> used by the pipeline)

Reply:
Thanks for the explanation. 

I thought if we use a resource like config map or secrets, the container can handle the case if they exist or not.
in this way if the cnbi is halted somewhere, we would have to handle those case, as it a get call.
maybe i m wrong here, so please do help me understand this.

another question would be in this way, at least we would need the exclusive name of the label/content right?
to gather the cnbi for appropriate requests?

Reply:
> Thanks for the explanation.
> 
> I thought if we use a resource like config map or secrets, the container can handle the case if they exist or not.

If we can handle it for a configMap, we can handle it for a cnbi, there is no difference (at least at the level of "does it exists or not ?")

>in this way if the cnbi is halted somewhere, we would have to handle those case, as it a get call. maybe i m wrong here, so please do help me understand this.

You mean in the case where the CNBi resource is deleted while the PipelineRun is in progress ? I see two possibility: either the PipelineRun is defined with ownerRef set to the CNBi resource (thus it's deleted) or we just let it error out. I would prefer the latter, since it preserves logs and that sort of stuff. (There is probably some TTL mechanism to garbage collect old pipelines somewhere in tekton...)

> another question would be in this way, at least we would need the exclusive name of the label/content right? to gather the cnbi for appropriate requests?

Reply:
if cnbi resource handle it, then great.

just another information share, there is no support for ttl on tekton side, at least on version 0.28.3 lower
the TTL were earlier managed by operator resources.

Reply:
moin all, looks like we have some design decisions in the comments. Can we refactor them to a ADR or at least the hackmd?

1. interface between CNBi Resource and PipelineRun - how are the packagelist parameters handed over?
1.1. if i/f is a configmap, the cnbi reconcile look should detect the non-existance of it and have a condition (that might go away again if CM is created) and only create a PipelineRun if all prerequisites are fulfilled?
2. ownerRef of PipelineRun - it should be owned by CNBi Resource and cascade on delete?

Reply:
Indeed, no *Runs cleanup on tekton... there is [TEP-0052](https://github.com/tektoncd/community/blob/main/teps/0052-tekton-results-automated-run-resource-cleanup.md) but it's not implemented.

@goern : Yes, we have some subject for the cnbi meeting today here. I'll put that on the agenda to reach a decision and write the ADR(s) when we're decided.